Fix GC reporting for slow tail call arguments of type `Span<T>`

kouvel opened the pull request:
Fixes https://github.com/dotnet/coreclr/issues/9032:
- Refactored by-ref-like method table walking to find offsets of by-ref pointers in siginfo.hpp/cpp
- Reused that for appending GC layout when creating the copy-args helper for a slow tail call

kouvel commented:
@jkotas, I didn't find the relevant code for generating the copy-args helper on arm64. Is it enabled there?

On arm the copy-args helper generation is not generating any GC layout. Are slow tail calls disabled there if the any of the args contain pointers? Wondering if something needs to be done now for by-ref-like args (maybe disable as well)? Also not sure how to properly disable slow tail calls in that case.

jkotas commented on src/vm/siginfo.hpp:
```diff
@@ -1188,7 +1188,43 @@ BOOL CompareTypeDefsForEquivalence(mdToken tk1, mdToken tk2, Module *pModule1, M
+        }
+
+        // TODO: GetApproxFieldTypeHandleThrowing may throw. This is a potential stress problem for fragile NGen of non-CoreLib
+        // assemblies. It won�t ever throw for CoreCLR with R2R. Figure out if anything needs to be done to deal with the
```
Nit: Weird character in won't

jkotas commented:
>  I didn't find the relevant code for generating the copy-args helper on arm64. Is it enabled there?

We have not done work to enable slow tail calls on arm64.

>  Are slow tail calls disabled there if any of the args contain pointers?

There are no structs passed by reference in arm calling convention. So this code is not needed there. It is only needed there are args passed by reference.

kouvel commented:
Moved `FindByRefPointerOffsetsInByRefLikeObject` to object.inl. It appears that putting the definition in siginfo.hpp created a circular dependency between definitions in class.h and siginfo.hpp. Fixed comment as well.

kouvel commented:
> diff.cs(22,31): error CS8137: Cannot define a class or member that utilizes tuples because the compiler required type 'System.Runtime.CompilerServices.TupleElementNamesAttribute' cannot be found. Are you missing a reference? [D:\j\workspace\x64_windows_n---616f30f2\jitutils\src\jit-diff\jit-diff.csproj]

Seems unrelated but happened twice in this PR and I don't see similar issues in other recent PRs. Any ideas?

kouvel commented:
@dotnet-bot test Ubuntu x64 Innerloop Formatting
test Windows_NT x64 Innerloop Formatting

AndyAyersMS commented:
@kouvel the formatter issue was fixed by #14785. Should pass on retry.

@dotnet-bot test Ubuntu x64 Innerloop Formatting 
@dotnet-bot test Windows_NT x64 Innerloop Formatting

kouvel commented:
Ok thanks, I'll fix the conflict and try again
